### ISSUE TITLE
'Fluent API' Recipe Builder syntax

### ISSUE THREAD

**Issue by webmilio** [tModLoader/tModLoader#847]
### Description
An extension coupled with a RecipeBuilder class, allowing for easier recipe building. The current implementation is in my library and works fine, although I assume it would need some adjusting to fit tML standards.

### What does this proposal attempt to solve or improve?

The way recipes are added right now is not exactly the most pleasant, albeit not atrocious. This syntax aims to solve the issue by providing some kind of 'natural language' implementation to recipes added in AddRecipes.

### Which (other) solutions should be considered?

I am unaware of any other solutions such as this one, other than Mirsario's.

### Examples
Here are two examples that show what the syntax could be.
The first example instantiates a new RecipeBuilder manually.

![image](https://user-images.githubusercontent.com/836345/80650836-b9eccd00-8a42-11ea-886a-3c9d2efeec88.png)

The second example uses both the extension method and multiple crafting stations

![image](https://user-images.githubusercontent.com/836345/80650927-eb659880-8a42-11ea-9508-773ea5331c87.png)

The current implementation has several other methods which I've not included in this issue, but are as follows:

- `SetResult`: calls `SetResult` for the recipe
- `SetResultStack`: changes only the amount produced, useful in case of multiple recipes producing the same item but not the same quantity
- Several instances of `New`, which take different parameters. `New` will automatically call `Finish` on an old recipe if its not been added to the game
- `Clear`: effectively calls `New` with the existing information. Not named `New` since it only works if you already have an existing product in the builder
- The different constructors for `RecipeBuilder` (which all follow the different `SetResult` for `ModRecipe`)

**Comment by Chicken-Bones** [comment]
I really like the syntax and concept.
My primary feedback is to make sure the system isn't too "overloaded". As per the zen of python "There should be one-- and preferably only one --obvious way to do it."

We want this to be _the one_ and _the obvious_ way
so `BuildRecipe()` would be the way to do it from within `AddRecipes()` for an item.

"New will automatically call Finish on an old recipe if its not been added to the game"
I don't like this, it's hidden, and could cause bugs with forgetting the last recipe in the chain. "Explicit is better than implicit."

I suggest replacing `RequiresSinglesOf` with just `Requires` and having a few overloads

perhaps the best way is to construct a recipebuilder with the target item and qty, and always have the last call be `At`. If no station is required just call `At(null)` or `Anywhere()`

**Comment by Jofairden** [comment]
First of all, thanks for your contribution / idea!
I agree with the things cb said, it should be simple and 'makes-sense' to follow.
Ditch RequiresSinglesOf, simply put the default stack on Requires as 1.

I think the following naming is fine and concise:
```
Ingredient
Requires
Produces
At
Finish/Complete/Add
```

Requires should be used for optional requirements such as water, honey and lava.

A couple helper methods may be useful in that regard, something like: 
```
RequiresLava()
RequiresWater()
RequiresHoney()
RequiresAlchemy()
``` 
etc.

It'd also be neat if there was a params version called Ingredients that you can use like so using value tuples:
```Ingredients((ItemID.Wood, 20), (ItemID.Sand, 10));```
For tiles:
```At(TileID.Workbenches, myTileId, otherTileID)```

We could probably use fluent API pattern in more places in TML.

**Comment by Trivaxy** [comment]
I'm all for this idea. I had my own RecipeBuilder sitting around for a while and I'd like to see one added to tML :)
Would RecipeBuilder instances be reusable, though? (don't see why not)

**Comment by webmilio** [comment]
They currently are reusable, yes!

**Comment by webmilio** [comment]
![image](https://user-images.githubusercontent.com/836345/80729595-780d6680-8ad6-11ea-8e72-796b6c2fe7aa.png)

Whats changed from the original post:

- The first new `Requires` has the following signature:
`int, int, int, int[]`, where the `int[]` is `params`. I did this instead of making multiple requires with increasing parameter count.
- The second is as Jofairden requested: `params (short itemId, int stack)[]`
- `New` no longer calls `Finish`.

I haven't removed `Clear` since making `At` empty the recipe every time feels too implicit, but I assume you wouldn't add a recipe to the game and then keep going from that same recipe. I'll add `Clear` to `At` if you guys think its a good idea.

**Comment by Jofairden** [comment]
One recipe builder instance should build one recipe.
1 <-> 1 relation

So ModRecipeBuilder.Build() should return a ModRecipe
If you want to make a new recipe, you then go again:
 `new ModRecipeBuilder().Requires(..)..At(...).Build();`

`Anywhere` sounds quite redundant since you don't require a tile, what does that even do?

**Comment by webmilio** [comment]
It replaces the `Finish` method, whereas you no longer have to call `Finish`. Instead, as soon as you call `At` or `Anywhere`, the recipe is assumed complete.

**Comment by Jofairden** [comment]
Yeah that's not how the builder pattern works
When you call 'Build()', that's the final step. Any other step is possible at any moment before building.
In case you're wondering, the fluent API pattern is almost always used alongside the builder pattern, and it makes perfect sense. In short, you should be able to construct the recipe in any order.

**Comment by Jofairden** [comment]
Assigning myself for now, we can probably make this really easily for future 1.4 TML version.

**Comment by webmilio** [comment]
I'll make a pull request for the existing implementation I have once everything works out!

**Comment by Mirsario** [comment]
Closing just like the PR was. It's all implemented for 1.4 now.
